### DISCUSSION TITLE
Claim 2021 Codeathon projects and request reviews here!

### DISCUSSION THREAD

**Discussion by anjohnson** [epics-base/epics-base#117] [Codeathon 2021]
We want to make sure that only one person (or one group working together) works on a particular project during the Codeathon, so nobody has their time wasted by duplication and so we can all see who's working on what. We also want participants to review  and test each other's work once it's ready for that.

Before you actually start working on any project (especially one taken from our [published lists](https://epics.anl.gov/meetings/codeathon-21/index.php#projects)), please check the other messages here and post a response in this discussion saying exactly what you're planning to work on. If two or more people pick the same project the first person to post here gets priority, so double-check after you've posted that nobody else got in first – maybe you could work together if that happens, it's worth asking, especially if the project involves any non-trivial development.

Similarly once you have pushed your changes and created a Pull Request or equivalent for a project and you'd like others to review, test and/or comment on it, please add a follow-up response here with a link to that PR, then look for other messages here about PR's that you might be able to review before you start on another project.

**Comment by tboegi** [comment]
That's a good point - is there a chance to collect a list of published lists here ?

**Reply by anjohnson** [reply]
While writing it I had a thought about adding them to the heading of this conversation, but that thought got crowded out by too many others and it eventually died a lonely and unfulfilled death at the time. As @ralphlange said (in a message that wasn't a direct reply to this question but could have been – just like the MS Teams chat UI it's easy to miss the "Write a reply" box under a comment and use the "Write a comment" box instead, which unfortunately risks losing the connection to the message it was replying to) the lists are available on the [Codeathon page](https://epics.anl.gov/meetings/codeathon-21/index.php#projects) but I will replicate them as suggested when I get a chance.

Sorry about the verbose language, my writing seems to be quite flowery today, I can't explain why...

**Reply by ralphlange** [reply]
_Meh!_
The "published lists" mentioned above are shown on the [Codeathon page](https://epics.anl.gov/meetings/codeathon-21/index.php) (below the time table).

**Reply by anjohnson** [reply]
Rather than duplicate the list pointers here, I added a link in the conversation heading above to the relevant section of the Codeathon webpage.

**Comment by 0ignacia0** [comment]
Hi all, I'm interested in contributing to the epicsTime rework project.
I'm familiar with the concept of C++ wrappers to provide an interface to C code. If I could be party to whatever discussion on project requirements/scope, I'd be happy to carry out any coding or testing tasks, given with a little direction. I've reviewed the related code, and I think I have an idea of what is required, but as I'm still a bit of a novice in deployed C++ development, I think it'd be beneficial to test this idea.

**Comment by tboegi** [comment]
I would like to have a look at "Accessing Record fields from asyn driver"
https://epics.anl.gov/tech-talk/2021/msg00404.php

**Reply by anjohnson** [reply]
@tboegi said: "In progress, a larger project that originally expected. Will continue with it after this week."

**Comment by dirk-zimoch** [comment]
I would like to claim „Read info items using a filter“ https://bugs.launchpad.net/epics-base/+bug/1902560

**Reply by simon-ess** [reply]
I would be happy to work with you on that; that was my plan as well.

**Reply by dirk-zimoch** [reply]
Repo is here: https://code.launchpad.net/~dirk.zimoch/epics-base/+git/epics-base/+ref/FilterForInfoFields

**Reply by dirk-zimoch** [reply]
Need to find out why it does not work with the latest dbChannel changes....

**Reply by bfrk** [reply]
Can you open a merge request so I can see the diffs and comment?

**Reply by dirk-zimoch** [reply]
Merge request: https://code.launchpad.net/~dirk.zimoch/epics-base/+git/epics-base/+merge/399414

**Comment by exzombie** [comment]
[Making TIME directly accessible](https://bugs.launchpad.net/epics-base/+bug/1182091) looks interesting, but there is not much information in the issue report. I'd like to discuss it.

**Reply by exzombie** [reply]
PR #126 is looking forward to comments

**Comment by zioven** [comment]
I can help with: [Add help for iocSh commands](https://github.com/epics-base/epics-base/wiki/code-a-thon-project-ideas#add-help-for-iocsh-commands)

**Reply by ebernalru** [reply]
Hi @zioven, I was thinking about this topic as well, let me know If I can be of any help with some commands or testing.

**Reply by zioven** [reply]
@ebernalru I have forked the base and I have created a "local" pull request on separate branch so I can track my progress

https://github.com/zioven/epics-base/pull/1

You can join in anytime, let just communicate through the following files:

I have added two text files in the `documentation` (which shall be removed before merge into main branch):
- first one (`list_of_iocshFundDef.txt`) contains all the all calls to `iocshFuncDef` in EPICS base
  - in this I also plan to track which function is already changed and which needs additional work (check the first few lines)
- second one (`list_of_documentedExamples.txt`) has a list of implemented help calls for reference...

To reduce work later on, I propose that the changes to the codebase and to the files above are different commits...

**Reply by oksanagit** [reply]
Hello,
I will be working on modules/database/src/ioc/misc/miscIocRegister.c . I will post links to my work in this discussion

**Reply by ebernalru** [reply]
Hi,
I'll focus on modules/database/src/ioc/db/dbIocRegister.c

**Reply by oksanagit** [reply]
My modifications are in the [local branch](https://github.com/oksanagit/epics-base/tree/help_iocsh_oksana). I will create a PR at the end of the week. So far help text is added for iocInit, iocBuild, iocRun, iocPause, coreRelease, system.

**Reply by oksanagit** [reply]
I am working on epics-base/modules/database/src/ioc/dbStatic/dbStaticIocRegister.c

**Reply by oksanagit** [reply]
Thanks @zioven for updating  list_of_iocshFundDef.txt . 
[oksanagit PR](https://github.com/epics-base/epics-base/pull/136) Done: dbDumpRecord, dbDumpMenu, dbDumpRecordType, dbDumpField, dbDumpDevice, dbDumpDriver, dbDumpRegistrar, dbDumpFunction, dbDumpVariable

**Reply by ebernalru** [reply]
Hi @anjohnson, yes, I'll keep working on this task after the codeathon if not done with it.

**Reply by zioven** [reply]
Pull request for my work: https://github.com/epics-base/epics-base/pull/142

Please review

**Reply by anjohnson** [reply]
#142 was merged.

**Comment by joaopaulosm** [comment]
I will continue working on [Make output records only write on change](https://bugs.launchpad.net/epics-base/+bug/1398215) as last Codeathon, expanding from _longout_ to other record types.

**Reply by joaopaulosm** [reply]
Pull-request #63 opened for the LONGOUT record. If approved, the same idea (OOPT field) can be used to extend the functionality to the other output records (bo, ao, mbbo, stringout, etc..)

**Comment by FreddieAkeroyd** [comment]
I am happy to work on compiler warnings https://bugs.launchpad.net/epics-base/+bug/1905159 if that would be useful

**Reply by gabrielfedel** [reply]
Freddie which compiler are you going to look into?

**Reply by FreddieAkeroyd** [reply]
I run on windows,  Visual Studio 2019

**Reply by gabrielfedel** [reply]
Good, I would like to take a look on this issue too then, but I will go with a newer gcc version (8.0.3 to start)

**Reply by FreddieAkeroyd** [reply]
PR at https://github.com/epics-base/epics-base/pull/129

**Reply by gabrielfedel** [reply]
PR at #130

**Reply by anjohnson** [reply]
#130 was merged.

**Comment by KathrynBaker** [comment]
Post for tracking [Add Doxygen annotations to C/C++ header files](https://github.com/epics-base/epics-base/wiki/code-a-thon-project-ideas#add-doxygen-annotations-to-cc-header-files)

**Reply by DominicOram** [reply]
I will be interested in dropping in and out of this today

**Reply by KathrynBaker** [reply]
I'm about to add some annotations to cvtFast

**Reply by KathrynBaker** [reply]
Adding annotations to freeList

**Reply by KathrynBaker** [reply]
Heading to the ThreadPool bits next

**Reply by anjohnson** [reply]
Hi Kathryn, the ThreadPool API hasn't really been used and might get dropped, so I suggest leaving that one for now.

**Reply by mp49** [reply]
I'd like to work on this. Should I look at a different directory other than libCom so we don't conflict?

**Reply by mdavidsaver** [reply]
I'll work on making a ~prioritized list of public, and not generated, headers.  My starting point will be in order of decreasing priority: libCom, libca, dbCore.

**Reply by KathrynBaker** [reply]
Thanks, a prioritised list would help, I'll revert back the minor changes I'd made to the ThreadPool bits, grand total of about 10 lines, mostly rearranging.

Where would you like mp94 and I to go most immediately?

**Reply by mdavidsaver** [reply]
https://github.com/epics-base/epics-base/wiki/Header-annotations-project

**Reply by anjohnson** [reply]
@KathrynBaker @DominicOram @mp49 
Those of you working on the Doxygen task might want to `git pull --rebase` some commits I just made to the 7.0 branch which changes the `Doxyfile` configuration to better match the one we used last year for libCom headers, and adds some code that annotates all the generated `*Record.h` files as well.

The new Doxyfile requires all annotated files to have a header something like this:
```C
/** @file filename.h
 * @brief Header file purpose/summary for the File List page
 *
 * More details about this file, optional
 */
```
You can use `\file` and `\brief` instead of course. One major difference to the output is that files which haven't been annotated aren't clickable on the File List page, so it's easy to see what has been done already and to find examples you can copy from.

Sorry that I didn't get this included before now. I don't plan to make any more changes like this...

**Reply by mp49** [reply]
I have added doxygen comments to errlog.h and to epicsStdlib.h
https://github.com/epics-base/epics-base/pull/127
https://github.com/epics-base/epics-base/pull/139
I would like to continue and do a few more files after this week.

**Reply by KathrynBaker** [reply]
I've put in a PR for the stdio.h I was looking at, unfortunately I've had to do less than I had hoped in the latter part of the week, apologies.

**Reply by anjohnson** [reply]
PRs #122 , #127 and #139 have been merged (after rebasing).

**Comment by DominicOram** [comment]
I've added https://github.com/epics-base/epics-base/wiki/code-a-thon-project-ideas#streamdevice-suppress-duplicate-timeout-messages as something I might do. @dirk-zimoch I would be interested to get your thoughts on it, does it sound like a sensible thing to do?

**Reply by dirk-zimoch** [reply]
Go ahed. I have just now pushed the latest changes from Freddie to allow switching off error colorization, Thus you may pull before starting.

**Reply by DominicOram** [reply]
Thanks, I think that as you said https://github.com/paulscherrerinstitute/StreamDevice/pull/39 is a good place to start, covers my usecase in a generic way. I'll fix the conflicts on it and tidy it up a bit

**Reply by DominicOram** [reply]
@marciodo would welcome your opinion on this too as you wrote the original PR

**Reply by marciodo** [reply]
Ok, I'd love to help.

**Reply by DominicOram** [reply]
I'm pretty far through this, hoping to finish tomorrow. If I don't I will carry it on next week.

**Reply by DominicOram** [reply]
PR at https://github.com/paulscherrerinstitute/StreamDevice/pull/68

**Comment by ralphlange** [comment]
I've added https://github.com/epics-base/epics-base/wiki/code-a-thon-project-ideas#generated-base-documentation-provide-versioned-content-on-readthedocs and started working on it. Has been on the list for a while, but I never had enough time to get it working.

**Comment by daykin** [comment]
I'd like to try to  [add SIMM=RAW to ao records.](https://bugs.launchpad.net/epics-base/+bug/1712363) This might be useful for some of our automated SNC state machine tests.

**Reply by daykin** [reply]
Update: Running into strange issue where intended behavior works in softIoc, but tests fail. Probably won't be fully resolved or generalized to all output records by end of codeathon.

**Reply by daykin** [reply]
Just noticed I never updated this. #144 Opened.

**Comment by jesusvasquez333** [comment]
I would like to work on the issue [pvtools argument handling](https://github.com/epics-base/epics-base/wiki/code-a-thon-project-ideas#pvtools-argument-handling) -> https://github.com/epics-base/pvAccessCPP/issues/168

**Reply by jesusvasquez333** [reply]
PR [#174](https://github.com/epics-base/pvAccessCPP/pull/174) opened for this issue.

**Reply by anjohnson** [reply]
This was merged.

**Comment by marciodo** [comment]
I'd like to claim "Teach iocsh about the current directory"

**Reply by anjohnson** [reply]
@marciodo submitted PR #123 for this.

**Reply by marciodo** [reply]
Update: I've submitted my last change. It is waiting for approval to merge.

**Comment by krzyslab** [comment]
I would like to look at  "#1902285 caput never uses ca_get_callback()"

**Reply by mdavidsaver** [reply]
cf. https://bugs.launchpad.net/epics-base/+bug/1902285

**Comment by hhslepicka** [comment]
I am claiming https://github.com/mdavidsaver/p4p/issues/60

**Reply by hhslepicka** [reply]
PR Submitted: https://github.com/mdavidsaver/p4p/pull/66

**Reply by anjohnson** [reply]
This was merged.

**Comment by ssaxenaANL** [comment]
I will be working on https://github.com/areaDetector/areaDetector/blob/master/docs/ToDo.md ADLambda documentation

**Reply by ssaxenaANL** [reply]
I have finished  https://github.com/areaDetector/areaDetector/blob/master/docs/ToDo.md ADLambda and ADPICam. Now, will be working on NDDriverStdArrays

**Reply by ssaxenaANL** [reply]
I have completed areaDetector drivers: NDDriverStdArrays and ADPvCAM. Now, I am working on ADQImaging.

**Reply by ssaxenaANL** [reply]
I have finished ADQImaging and all of the documentation I did during the week are merged now. I will keep working on ADPcoWin.

**Comment by kukhee** [comment]
I am interested in the "periodic SCAN tasks use clock_nanosleep()"  in Wiki page:
https://github.com/epics-base/epics-base/wiki/code-a-thon-project-ideas

As Mike mentioned in the comments: we need to test carefully, and we also need to check up the compatibility, some of OS (ex, RTEMS) may not support the monotonic clock option. I would like to look at more details.

**Reply by kukhee** [reply]
I have investigated the issue for using clcok_nanosleep() with monotonic clock option for periodic SCAN task,
and summarized my investigation in the following link.

https://1drv.ms/b/s!AsH0Qpew48nUh4ZubD6mF6VgvrLXtg?e=mZle6V

I need to go back to my work, so I could not implement it and test it.
But, if you have questions and discussions about the issue, I am welcome anytime.
Please, do not hesitate to contact me about the issue.
Thank you.

**Comment by rerpha** [comment]
May have a go at https://github.com/epics-base/epics-base/issues/80 , although I would like to join someone and do some pair programming as I haven't worked on the internals of epics before.

**Reply by rerpha** [reply]
PR: https://github.com/epics-base/epics-base/pull/125

**Reply by anjohnson** [reply]
Merged, thanks! 🎉

**Comment by jesusvasquez333** [comment]
I'm going to take a look at [Improve the CAS error message generated when an IOC gets port-scanned](https://github.com/epics-base/epics-base/wiki/code-a-thon-project-ideas#improve-the-cas-error-message-generated-when-an-ioc-gets-port-scanned)

**Reply by jesusvasquez333** [reply]
I translate that issue into GitHub Issue https://github.com/epics-base/epics-base/issues/128

**Reply by jesusvasquez333** [reply]
PR https://github.com/epics-base/epics-base/pull/141 opened for this issue.

While working on this issue: 
- I did some minor fixes to the CA protocol documentation, and opened this PR: https://github.com/epics-docs/epics-docs/pull/3 
- I might found a error in the CA protocol documentation, so I opened this issue: https://github.com/epics-docs/epics-docs/issues/2

**Comment by marciodo** [comment]
I'll give it a try at Stream Device [Add a new checksum method](https://github.com/paulscherrerinstitute/StreamDevice/issues/60#).

**Reply by marciodo** [reply]
Update: I hope I can finish this before the end of the Codeathon.

**Reply by marciodo** [reply]
Stream Device PR [#69](https://github.com/paulscherrerinstitute/StreamDevice/pull/69) created.

**Comment by mdavidsaver** [comment]
While maybe not a "proper" codeathon project since I did the work 3 weeks ago, I'd be interested in having some help looking over #113 (errlog rewrite w/ color removal).

**Comment by anjohnson** [comment]
To everyone who has "claimed" a task here but hasn't posted about it recently: Please give us an update (by replying to your original claim message in [this GitHub discussion](https://github.com/epics-base/epics-base/discussions/117)) to let us know if you are still working on that claim and if so whether you plan to continue with it after this week.

We will be holding wrap-up sessions on the Zoom channel on Friday; for the UK/European participants at 15:00 UTC, 16:00 CET and for the American participants at 1pm PST, 3pm CST, 4pm EST. If you can attend one (or both) of those sessions we would like to hear about your progress on your tasks, what you learned, and how we can make future virtual Codeathons better and easier.

Thank-you for attending and being willing to help us to improve EPICS. If you enjoyed working on it and want to continue, feel free to talk to any of the Core Developers group members about taking on projects outside of the context of a Codeathon, we can always use more help!

- Andrew

**Comment by simon-ess** [comment]
If someone wants to review my syncing of softMain.cpp from EPICS base and from pva2pva: https://github.com/epics-base/pva2pva/pull/40

**Reply by anjohnson** [reply]
This has been merged.
